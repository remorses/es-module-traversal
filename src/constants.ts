import os from 'os'
export const PACKAGE_NAME = 'es-module-traversal'

export const MAX_IO_OPS: number = os.cpus().length * 4

export const JS_EXTENSIONS = new Set([
    '.js',
    '.jsx',
    '.ts',
    '.tsx',
    '.mjs',
    '.cjs',
    '.vue',
    '.scss',
    '.css',
])
