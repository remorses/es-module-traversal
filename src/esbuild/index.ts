export { traverseWithEsbuild, metaToTraversalResult } from './traverseEsbuild'
