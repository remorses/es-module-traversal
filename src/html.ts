import posthtml, { Node, Plugin } from 'posthtml'
import { isRelative } from './support'

export async function getHtmlScriptsUrls(html: string) {
    const urls: string[] = []
    const transformer = posthtml([
        async (tree) => {
            tree.walk((node) => {
                if (
                    node.tag === 'script' &&
                    node.attrs['type'] === 'module' &&
                    node.attrs['src'] &&
                    isRelative(node.attrs['src'])
                ) {
                    urls.push(node.attrs['src'])
                }
                return node
            })
        },
    ])
    const result = await transformer.process(html)
    return urls.filter(Boolean)
}
