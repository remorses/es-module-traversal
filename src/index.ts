import { promises as fsp } from 'fs'
import fs from 'fs'
import path from 'path'
import resolve from 'enhanced-resolve'
import { init, parse } from 'es-module-lexer'
import { debug } from './constants'
import isBuiltin from 'is-builtin-module'

const defaultResolver = resolve.create.sync({
    extensions: ['.js', '.jsx', '.ts', '.tsx', '.mjs'],
})

function isRelative(x: string) {
    return x.startsWith('.') || x.startsWith('/')
}

export type ResultType = {
    importPath: string
    resolved?: string
    importer: string
}

export function* walkEsModulesGenerator(
    entryPoint,
    resolver = defaultResolver,
): Generator<ResultType> {
    const content = fs.readFileSync(entryPoint).toString()
    const [imports, exports] = parse(content)
    for (const { s, e, d } of imports) {
        let importPath = content.slice(s, e).trim()
        const isDynamicImport = d > -1
        if (isDynamicImport) {
            try {
                importPath = eval(importPath)
            } catch {
                continue
            }
        }
        debug(importPath)
        if (isBuiltin(importPath)) {
            continue
        }
        const resolved =
            resolver(path.dirname(entryPoint), importPath) || undefined
        yield { importPath, resolved, importer: entryPoint }
        if (isRelative(importPath)) {
            yield* walkEsModulesGenerator(resolved, resolver)
        }
    }
}

export function walkEsModulesSync({ entryPoint, resolver = defaultResolver }) {
    return [...walkEsModulesGenerator(entryPoint, resolver)]
}
