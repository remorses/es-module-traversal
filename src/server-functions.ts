import fetch from 'node-fetch'
import path from 'path'
import slash from 'slash'
import { URL } from 'url'
import { defaultRead, defaultResolver, isRelative } from '.'
import { PACKAGE_NAME } from './constants'
import { cleanUrl, debug, isUrl, readFromDisk } from './support'

export async function readFromUrlOrPath(url: string, importer?: string) {
    let content = ''
    if (!isUrl(url)) {
        content = await readFromDisk(url)
    } else {
        const isHtml =
            path.extname(cleanUrl(url).split('/').reverse()[0]) === '.html'
        const Accept =  isHtml ? 'text/html' : '*/*'
        const res = await fetch(url, {
            headers: {
                ...(importer ? { Referer: importer } : {}),
                Accept,
                'User-Agent': PACKAGE_NAME,
            },
        })
        if (!res.ok) {
            throw new Error(
                `Cannot fetch '${url}', referer is '${importer}', content type is '${Accept}': ${
                    res.statusText
                } ${await res.text().catch(() => '')}`,
            )
        }
        content = await res.text()
    }

    return content
}

/*
 * Create a resolver function that support urls and paths
 * It assumes a static server behind `baseUrl` with `root` folder mounted at /
 */
// TODO rename to browserEmulator
export const urlResolver = ({
    root,
    baseUrl,
}: {
    root: string
    baseUrl: string
}) => {
    // resolves relative paths to absolute paths and adds the root baseUrl
    return function resolveUrlOrPath(ctx: string, importPath: string = '') {
        debug(`resolveUrlOrPath from '${ctx}' to '${importPath}'`)
        let importerDirectory = isUrl(ctx)
            ? urlToRelativePath(ctx)
            : slash(path.relative(root, ctx)) // TODO assumes importer files paths are always absolute
        if (!isRelative(importPath)) {
            // console.log('defaultResolver from', importerDirectory, importPath)
            return defaultResolver(importerDirectory, importPath)
        }
        // console.log({ importerDirectory, importPath, root })
        if (importPath.startsWith('.')) {
            // resolve relative to the importer file
            importPath = path.join(importerDirectory, importPath)
        } else if (importPath.startsWith('/')) {
            // new URL adds // if import path starts with /
            importPath = importPath.slice(1)
        }
        // console.log(importPath)
        // console.log({ importPath, pathname: importerDirectory })
        const url = new URL(importPath, baseUrl).toString()
        // debug({ from: importPath, to: url, ctx })
        return url
    }
}

function urlToRelativePath(ctx) {
    let pathname = new URL(ctx).pathname
    pathname = pathname.startsWith('/') ? pathname.slice(1) : pathname
    return pathname
}
