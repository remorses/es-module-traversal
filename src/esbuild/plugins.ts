import { Plugin } from 'esbuild'
import builtins from 'builtin-modules'

// resolve node builtins

// mark node modules as external

export function StopTraversingPlugin({ stopTraversing }): Plugin {
    return {
        name: 'stop-traversing',
        setup: function setup({ onLoad, onResolve }) {
            onResolve({ filter: /.*/ }, (args) => {
                console.log({ args })
                const external = stopTraversing(
                    args.resolveDir, // TODO replace resolveDir with the complete path to file
                    args.importer,
                )
                return {
                    external,
                }
            })
        },
    }
}

export function CustomResolverPlugin({ resolver }): Plugin {
    const builtinsSet = new Set(builtins)
    return {
        name: 'stop-traversing',
        setup: function setup({ onLoad, onResolve }) {
            onResolve({ filter: /.*/ }, (args) => {
                if (builtinsSet.has(args.path)) {
                    return {
                        external: true,
                    }
                }
                const resolved = resolver(args.importer, args.path)

                // console.log({ resolved })
                return {
                    external: !resolved,
                    path: resolved,
                }
            })
        },
    }
}
