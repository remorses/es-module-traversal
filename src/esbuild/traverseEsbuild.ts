import {
    NodeResolvePlugin
} from '@esbuild-plugins/all'
import deepmerge from 'deepmerge'
import { build, BuildOptions, Metadata, Plugin } from 'esbuild'
import { promises as fsp } from 'fs'
import fsx from 'fs-extra'
import os from 'os'
import path from 'path'
import slash from 'slash'
import { JS_EXTENSIONS } from '../constants'
import {
    defaultResolver,
    flatten,
    isRunningWithYarnPnp,
    unique
} from '../support'
import { TraversalResultType } from '../types'

// TODO support tsconfig paths,
// TODO make any non js module external
// use the right resolver in esbuild
// maybe can add support for vue and other stuff by loading a vue file to a js file? i will let people pas the loaders for esbuild directly, this way they can pass them if they want
// mark as external modules that make stopTraversing true

type Args = {
    entryPoints: string[]
    esbuildOptions?: Partial<BuildOptions>
    resolver?: (cwd: string, id: string) => string
    stopTraversing?: (resolvedPath: string) => boolean
}

// resolver = defaultResolver,
// onEntry,
// stopTraversing,
// ignore = [],
// readFile = defaultRead,
export async function traverseWithEsbuild({
    entryPoints,
    resolver = defaultResolver,
    esbuildOptions = { plugins: [] },
    stopTraversing,
}: Args): Promise<TraversalResultType[]> {
    const destLoc = path.resolve(
        await fsp.mkdtemp(path.join(os.tmpdir(), 'dest')),
    )

    entryPoints = entryPoints.map((x) => path.resolve(x))

    try {
        const metafile = path.join(destLoc, 'meta.json')

        const esbuildCwd = process.cwd()
        await build(
            deepmerge(
                {
                    // splitting: true, // needed to dedupe modules
                    // external: externalPackages,

                    minifyIdentifiers: false,
                    minifySyntax: false,
                    minifyWhitespace: false,
                    mainFields: ['module', 'browser', 'main'],
                    sourcemap: false,
                    define: {
                        'process.env.NODE_ENV': JSON.stringify('dev'),
                        global: 'window',
                        // ...generateEnvReplacements(env),
                    },
                    inject: [
                        // require.resolve(
                        //     '@esbuild-plugins/node-globals-polyfill/process.js',
                        // ),
                    ],
                    // tsconfig: ,
                    loader: {
                        '.js': 'jsx',
                    },
                    plugins: [
                        ExternalButInMetafile(),
                        // NodeModulesPolyfillPlugin({ fs: true, crypto: true }), // TODO enable if in browser?
                        NodeResolvePlugin({
                            external: function external(resolved) {
                                if (
                                    stopTraversing &&
                                    stopTraversing(resolved)
                                ) {
                                    return {
                                        namespace: externalNamespace,
                                        path: resolved,
                                    }
                                }
                                return false
                            },
                            onUnresolved: () => {
                                return {
                                    external: true,
                                }
                            },
                            resolveOptions: {
                                preserveSymlinks: isRunningWithYarnPnp || false,
                                extensions: [...JS_EXTENSIONS],
                            },
                        }),
                    ].filter(Boolean),
                    bundle: true,
                    platform: 'node',
                    format: 'esm',
                    write: true,
                    entryPoints,
                    outdir: destLoc,
                    minify: false,
                    logLevel: 'info',
                    metafile,
                } as BuildOptions,
                esbuildOptions,
            ),
        )

        const meta: Metadata = JSON.parse(
            await (await fsp.readFile(metafile)).toString(),
        )
        // console.log(await (await fsp.readFile(metafile)).toString())

        const res = flatten(
            entryPoints.map((entry) => {
                return metaToTraversalResult({ meta, entry, esbuildCwd })
            }),
        ).map((x) => {
            return {
                ...x,
                resolvedImportPath: removeColon(x.resolvedImportPath),
                importer: removeColon(x.importer),
            }
        })
        return res
    } finally {
        await fsx.remove(destLoc)
    }
}

const externalNamespace = 'external-but-keep-in-metafile'
function ExternalButInMetafile(): Plugin {
    return {
        name: externalNamespace,
        setup(build) {
            const externalModule = 'externalModuleXXX'
            build.onResolve({ filter: new RegExp(externalModule) }, (args) => {
                if (args.path !== externalModule) {
                    return
                }
                return {
                    external: true,
                }
            })
            build.onLoad(
                {
                    filter: /.*/,
                    namespace: 'external-but-keep-in-metafile',
                },
                (args) => {
                    const contents = `export * from '${externalModule}'`
                    return { contents, loader: 'js' }
                },
            )
        },
    }
}

function removeColon(input: string) {
    const index = input.indexOf(':')
    if (index === -1) {
        return input
    }
    const clean = input.slice(index + 1)
    return clean
}

// TODO i cannot get the import paths with esbuild, output will only be with resolvedPath and importer

export function metaToTraversalResult({
    meta,
    entry,
    esbuildCwd = '.',
}: {
    meta: Metadata
    esbuildCwd: string
    entry: string
}): TraversalResultType[] {
    if (!path.isAbsolute(esbuildCwd)) {
        throw new Error('esbuildCwd must be an absolute path')
    }
    if (!path.isAbsolute(entry)) {
        throw new Error('entry must be an absolute path')
    }
    const alreadyProcessed = new Set<string>()
    let toProcess = [slash(path.relative(esbuildCwd, entry))]
    let result: TraversalResultType[] = []
    while (toProcess.length) {
        const newImports = flatten(
            toProcess.map((newEntry) => {
                if (alreadyProcessed.has(newEntry)) {
                    return
                }
                alreadyProcessed.add(newEntry)
                // newEntry = path.posix.normalize(newEntry) // TODO does esbuild always use posix?
                const input = meta.inputs[newEntry]
                if (!input) {
                    throw new Error(
                        `entry ${newEntry} is not present in esbuild metafile`,
                    )
                }
                const currentImports = input.imports.map((x) => x.path)
                // newImports.push(...currentImports)
                result.push(
                    ...currentImports.map(
                        (x): TraversalResultType => {
                            return {
                                importPath: '',
                                importer: path.join(esbuildCwd, newEntry),
                                resolvedImportPath: path.join(esbuildCwd, x),
                            }
                        },
                    ),
                )
                return currentImports
            }),
        ).filter(Boolean)
        toProcess = newImports
    }
    return unique(result, (x) => x.importPath + x.resolvedImportPath)
    // find the right output getting the key of the right output.inputs == input
    // get the imports of the inputs.[entry].imports and attach them the importer
    // do the same with the imports just found
    // return the list of input files
}
