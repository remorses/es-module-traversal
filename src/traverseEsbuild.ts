import { build, Metadata } from 'esbuild'
import path from 'path'
import { promises as fsp } from 'fs'
import { defaultResolver, defaultReadFile } from '.'
import { TraverseArgs, TraversalResultType } from './types'
import { flatten } from './support'

// TODO support tsconfig paths,
// TODO make any non js module external
// use the right resolver in esbuild
// maybe can add support for vue and other stuff by loading a vue file to a js file? i will let people pas the loaders for esbuild directly, this way they can pass them if they want
// mark as external modules that make stopTraversing true
export async function traverseWithEsbuild({
    entryPoints,
    resolver = defaultResolver,
    onFile,
    stopTraversing,
    ignore = [],
    readFile = defaultReadFile,
}: TraverseArgs): Promise<TraversalResultType[]> {
    const destLoc = await fsp.mkdtemp('dest')

    const metafile = path.join(destLoc, 'meta.json')

    await build({
        splitting: true, // needed to dedupe modules
        // external: externalPackages,
        minifyIdentifiers: false,
        minifySyntax: false,
        minifyWhitespace: false,
        mainFields: ['module', 'browser', 'main'].filter(Boolean),
        // sourcemap: 'inline', // TODO sourcemaps panics and gives a lot of CPU load
        define: {
            'process.env.NODE_ENV': JSON.stringify('dev'),
            global: 'window',
            // TODO defined to make any package work
            // ...generateEnvReplacements(env),
        },
        // TODO inject polyfills for runtime globals like process, ...etc
        // TODO allow importing from node builtins when using allowNodeImports
        // TODO add plugin for pnp resolution
        // tsconfig: ,
        bundle: true,
        format: 'esm',
        write: true,
        entryPoints,
        outdir: destLoc,
        minify: false,
        logLevel: 'info',
        metafile,
    })

    const meta: Metadata = JSON.parse(
        await (await fsp.readFile(metafile)).toString(),
    )

    const res = flatten(
        entryPoints.map((entry) => {
            return metaToTraversalResult({ meta, entry })
        }),
    )
    return res
}

// TODO i cannot get the import paths with esbuild, output will only be with resolvedPath and importer

export function metaToTraversalResult({
    meta,
    entry,
}: {
    meta: Metadata
    entry: string
}): TraversalResultType[] {
    let imports = [entry] // TODO esbuild has relative imports to the cwd? i have to resolve them first
    let result: TraversalResultType[] = []
    while (imports.length) {
        const newImports = []
        imports.forEach((newEntry) => {
            const input = meta.inputs[newEntry]
            if (!input) {
                throw new Error(
                    `entry ${newEntry} is not present in esbuild metafile`,
                )
            }
            const currentImports = input.imports.map((x) => x.path)
            newImports.push(...currentImports)
            result.push(
                ...currentImports.map(
                    (x): TraversalResultType => {
                        return {
                            importPath: '',
                            importer: newEntry,
                            resolvedImportPath: x,
                        }
                    },
                ),
            )
        })
        imports = newImports
    }
    return result
    // find the right output getting the key of the right output.inputs == input
    // get the imports of the inputs.[entry].imports and attach them the importer
    // do the same with the imports just found
    // return the list of input files
}
