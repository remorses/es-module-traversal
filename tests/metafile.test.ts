import { Metadata } from 'esbuild'
import { metaToTraversalResult } from '../src/traverseEsbuild'

it('metaToTraversalResult', async () => {
    const meta: Metadata = {
        outputs: {},
        inputs: {
            'entry': {
                bytes: 0,
                imports: [{ path: 'some-file' }],
            },
            'some-file': {
                bytes: 0,
                imports: [{ path: 'other-file' }],
            },
            'other-file': {
                bytes: 0,
                imports: [],
            },
        },
    }
    const res = metaToTraversalResult({
        meta: meta as any,
        entry: '/usr/someFolder/entry',
        esbuildCwd: '/usr/someFolder',
    })
    // console.log(res)
    expect(res).toMatchSnapshot('simple metaToTraversalResult')
})
