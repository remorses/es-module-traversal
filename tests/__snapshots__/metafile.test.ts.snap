// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metaToTraversalResult: simple metaToTraversalResult 1`] = `
Array [
  Object {
    "importPath": "",
    "importer": "./entry",
    "resolvedImportPath": "./some-file",
  },
  Object {
    "importPath": "",
    "importer": "./some-file",
    "resolvedImportPath": "./other-file",
  },
]
`;
